bee-tangle/src/tangle.rs

```diff
@@ -1,759 +1,133 @@
+            Entry::Vacant(entry) => {
+                let vtx = Vertex::new(data, metadata);
+                let txref = vtx.get_data().clone();
+                entry.insert(vtx);
+                Some(txref)
+            }
         }
     }
 
-    /// Returns a [`VertexRef`] linked to the specified milestone, if it's available in the local Tangle.
-    pub fn get_latest_milestone(&'static self) -> Option<TransactionRef> {
-        todo!("get the last milestone index, get the transaction hash from it, and query the Tangle for it")
-    }
-
-    /// Returns the hash of a milestone.
-    pub fn get_milestone_hash(&'static self, index: MilestoneIndex) -> Option<Hash> {
-        match self.milestones.get(&index) {
-            None => None,
-            Some(v) => Some(*v),
+    #[inline]
+    fn add_child(&self, parent: TxHash, child: TxHash) {
+        match self.children.entry(parent) {
+            Entry::Occupied(mut entry) => {
+                let children = entry.get_mut();
+                children.push(child);
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(vec![parent]);
+            }
         }
     }
 
-    /// Returns whether the milestone index maps to a know milestone hash.
-    pub fn contains_milestone(&'static self, index: MilestoneIndex) -> bool {
-        self.milestones.contains_key(&index)
-    }
-
-    /// Retreives the solid milestone index.
-    pub fn get_solid_milestone_index(&'static self) -> MilestoneIndex {
-        self.solid_milestone_index.load(Ordering::Relaxed).into()
-    }
-
-    /// Updates the solid milestone index to `new_index`.
-    pub fn update_solid_milestone_index(&'static self, new_index: MilestoneIndex) {
-        self.solid_milestone_index.store(*new_index, Ordering::Relaxed);
-    }
-
-    /// Retreives the snapshot milestone index.
-    pub fn get_snapshot_milestone_index(&'static self) -> MilestoneIndex {
-        self.snapshot_milestone_index.load(Ordering::Relaxed).into()
-    }
-
-    /// Updates the snapshot milestone index to `new_index`.
-    pub fn update_snapshot_milestone_index(&'static self, new_index: MilestoneIndex) {
-        self.snapshot_milestone_index.store(*new_index, Ordering::Relaxed);
-    }
-
-    /// Retreives the last milestone index.
-    pub fn get_last_milestone_index(&'static self) -> MilestoneIndex {
-        self.last_milestone_index.load(Ordering::Relaxed).into()
+    /// Get the data of a vertex associated with the given `hash`.
+    pub fn get_data(&self, hash: &TxHash) -> Option<TxRef> {
+        self.vertices.get(hash).map(|vtx| vtx.value().get_data().clone())
     }
 
-    /// Updates the last milestone index to `new_index`.
-    pub fn update_last_milestone_index(&'static self, new_index: MilestoneIndex) {
-        self.last_milestone_index.store(*new_index, Ordering::Relaxed);
+    /// Get the metadata of a vertex associated with the given `hash`.
+    pub fn get_metadata(&self, hash: &TxHash) -> Option<T> {
```

Comment:
So this is copying the metadata ?

Reply:
yea. It probably doesn't have to stay this way, but currently it is required to be a `Copy` type. Until our metadata increases to more than 8 bytes this should be fine.